nearcore/src/cold_storage.rs

```diff
@@ -0,0 +1,183 @@
+    let cold_head_hash = match cold_head_hash {
+        Some(cold_head_hash) => cold_head_hash,
+        None => return Err(anyhow::anyhow!("Failed to read the cold head hash")),
+    };
+
+    // The previous block is the cold head so we can use it to get epoch id.
+    let epoch_id = &runtime.get_epoch_id_from_prev_block(&cold_head_hash)?;
+    let shard_layout = runtime.get_shard_layout(epoch_id)?;
+
+    update_cold_db(cold_db, hot_store, &shard_layout, &next_height)?;
+
+    update_cold_head(cold_db, hot_store, &next_height)?;
+
+    if next_height >= hot_final_head_height {
+        Ok(ColdStoreLoopStatus::Sleep)
+    } else {
+        Ok(ColdStoreLoopStatus::Continue)
+    }
+}
+
+// This method will copy data from hot storage to cold storage in a loop.
+// It will try to copy blocks as fast as possible up until cold head = final head.
+// Once the cold head reaches the final head it will sleep for one second before
+// trying to copy data at the next height.
+// TODO clean up the interface, currently we need to pass hot store, cold store and
+// cold_db which is redundant.
+fn cold_store_loop(
+    keep_going: Arc<AtomicBool>,
+    hot_store: Store,
+    cold_store: Store,
+    cold_db: Arc<ColdDB>,
+    genesis_height: BlockHeight,
+    runtime: Arc<NightshadeRuntime>,
+) {
+    tracing::info!(target: "cold_store", "starting cold store loop");
+
+    // TODO - is there a better way to schedule a function to run periodically?
+    // NOTE - sleeping here, if spawned in rayon, would likely hog the thread forever
+    // which is why this method is spawned in rust native thread. The hope is that
+    // the OS will correctly handle this thread sleeping and return the CPU time to
+    // other threads.
+    loop {
+        if !keep_going.load(std::sync::atomic::Ordering::SeqCst) {
+            tracing::debug!(target: "cold_store", "stopping the cold store loop");
+            break;
+        }
+        let status = cold_store_copy(&hot_store, &cold_store, &cold_db, genesis_height, &runtime);
+
+        match status {
+            Err(err) => {
+                tracing::error!(target:"cold_store", "cold_store_copy failed with error : {err}");
+                std::thread::sleep(std::time::Duration::from_secs(1));
+            }
+            Ok(ColdStoreLoopStatus::Sleep) => {
+                std::thread::sleep(std::time::Duration::from_secs(1));
+            }
+            Ok(ColdStoreLoopStatus::Continue) => {
+                continue;
+            }
+        }
+    }
+}
+
+/// Spawns the cold store loop in a background thread and returns ColdStoreLoopHandle.
+/// If cold store is not configured it does nothing and returns None.
+pub fn spawn_cold_store_loop(
+    config: &NearConfig,
+    storage: &NodeStorage,
+    runtime: Arc<NightshadeRuntime>,
+) -> anyhow::Result<Option<ColdStoreLoopHandle>> {
+    tracing::info!("tracing info spawn cold store loop");
```

Comment:
you probably want to add a 'target' for tracing

Reply:
nit: a little bit confusing tracing levels.. here you say 'spawn' @INFO, but couple lines later you say 'not spawning' on DEBUG...

so many people would just see 'spawn' in their logs, and will be confused.

Maybe move it to line 165?